### ISSUE TITLE
`validate()` does not validate non-named elements

### ISSUE THREAD

**Issue by laco0416** [PolymerElements/iron-form#48]
I think that `validate()` should validate `type` or `required` of `<input>`.

 why is this valid?

```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <script src="bower_components/webcomponentsjs/webcomponents-lite.js"></script>
  <link rel="import" href="bower_components/iron-form/iron-form.html">
  <link rel="import" href="bower_components/iron-input/iron-input.html">

  <script>
    function clickHandler(event) {
      var form = Polymer.dom(event).localTarget.parentElement;
      if (form.validate()) {
        console.log("OK!");
      } else {
        console.log("Invalid!");
      }
    }
  </script>
</head>
<body>
  <form is="iron-form" method="get" action="/">
    <label>iron-input[type=email][required]</label><br/>
    <input is="iron-input" type="email" required/><br/>
    <button onclick="clickHandler(event)">Submit</button>
  </form>
</body>
</html>
```

![image](https://cloud.githubusercontent.com/assets/1529180/9524963/ae7e178e-4d1c-11e5-8bf9-dde84022fe95.png)

![image](https://cloud.githubusercontent.com/assets/1529180/9524994/d1f7519e-4d1c-11e5-98a4-f1dabb058327.png)

Thanks.

**Comment by laco0416** [comment]
I think the reason is because `iron-form-element-register` event has fired in `IronFormElementBehavior.attached()` only. 

https://github.com/PolymerElements/iron-form-element-behavior/blob/master/iron-form-element-behavior.html#L75

**Comment by laco0416** [comment]
the reason is here! 
http://jsbin.com/yusiciveno/4/edit?html,js,output

named elements are always validated, but non-named aren't.

**Comment by arthurevans** [comment]
I may be misreading the spec, but AFAIKT, the vanilla `<form>` element should only be validating "submittable elements":

https://html.spec.whatwg.org/multipage/forms.html#barred-from-constraint-validation

Submittable elements are the ones used for "creating the form data set":

https://html.spec.whatwg.org/multipage/forms.html#category-submit

Which omits almost all unnamed elements, except for one weird exception having to do with image buttons:

https://html.spec.whatwg.org/multipage/forms.html#constructing-form-data-set

Basically: only named inputs get _submitted_, so only named inputs are validated. This seems correct, but it should be noted in the docs.

**Comment by laco0416** [comment]
@arthurevans thanks to your explain!
does this mean that Chrome validation is out of spec?
i think browser and polymer should have similar validations.

**Comment by notwaldorf** [comment]
I've been doing more thinking about this. The native form implementation (example: http://jsbin.com/benome/edit?html,output) will validate elements even if they're not named. I think this line: https://github.com/PolymerElements/iron-form/blob/master/iron-form.html#L329 should probably drop the `if (required)` statement.

@laco0416 Let me know if you're interested in putting together a PR to fix this. Otherwise, I can take a look at it.

**Comment by laco0416** [comment]
@notwaldorf Hi, notwaldorf. Thanks to your recall.

>  I think this line: https://github.com/PolymerElements/iron-form/blob/master/iron-form.html#L329 should probably drop the if (required) statement.

I have other view.  I think that the problem line is here : https://github.com/PolymerElements/iron-form/blob/master/iron-form.html#L341
I think it should drop `&& el.name` statement.

**Comment by notwaldorf** [comment]
Oh yes, you're completely right. I should stop triaging issues at midnight :)

**Comment by laco0416** [comment]
uh-oh, I see. should I send a patch for this soon?

**Comment by notwaldorf** [comment]
That would be amazing (even more so if it has a test for this behaviour)! Thanks so much! 🍰

**Comment by laco0416** [comment]
@notwaldorf I sent a PR just now! please review it. 
https://github.com/PolymerElements/iron-form/pull/99
